service/history/replication/executable_activity_state_task.go

```diff
@@ -209,3 +240,54 @@ func (e *ExecutableActivityStateTask) MarkPoisonPill() error {
 
 	return writeTaskToDLQ(ctx, e.DLQWriter, shardContext, e.SourceClusterName(), replicationTaskInfo)
 }
+
+func (e *ExecutableActivityStateTask) BatchWith(incomingTask BatchableTask) (TrackableExecutableTask, bool) {
+	if !e.batchable || !incomingTask.CanBatch() {
```

Comment:
You're accessing this outside of the lock here but only setting it inside the lock.

This is a race condition and could lead to undesireable behavior

Reply:
Follow up PR to remove the lock: https://github.com/temporalio/temporal/pull/5280 please take a look when you got time :)